sensor: overflow in sensor_value from double

5frank opened the pull request:
Fixes #39176

Signed-off-by: Simon Frank <simon.frank@lohmega.com>

MaureenHelm reviewed (commented):
Thanks for the fix!

Please review and fix the compliance check failure

avisconti reviewed (commented):
why not squashing the two commits?

Commit:
include: sensor: sensor_value from double overflow

sensor_value_from_double had a early overflow when converting the
fractional part (val2).  This occured when input was more then
2147.493647 (inp >= INT32_MAX/1000000.0 + 0.01).

return value -ERANGE as this is what errno is set to by `strtod` and
similar posix functions.

fixes issue #39176

Signed-off-by: Simon Frank <simon.frank@lohmega.com>

teburd reviewed (changes requested):
ina219, stm32_temp, and i3g4250 use this call and are not expecting a fallible call. Those drivers should be updated to check for the error and return errors themselves or cast the return to (void) to annotate the call as ignoring the errno if its impossible to fail.

5frank commented:
Not sure why Documentation build was cancelled? I doesn't look like it relates to any of the changes in this PR?

gmarull commented:
> Not sure why Documentation build was cancelled? I doesn't look like it relates to any of the changes in this PR?

Just rebase, the doc issue was fixed a few weeks ago.

dleach02 commented:
@5frank, can you resolve the conflict?

henrikbrixandersen commented:
@5frank Please rebase on `main`, fix the merge conflicts, and we'll see if that fixes the CI issue as well.

ycsin commented on include/drivers/sensor.h:
```diff
@@ -666,11 +666,24 @@ static inline double sensor_value_to_double(const struct sensor_value *val)
+
+	double val2 = (inp - (int32_t)inp) * 1000000.0;
+
+	if (val2 < INT32_MIN || val2 > INT32_MAX) {
```
I wonder if the `val2` (same goes for the `inp`) above is treated as an `int` or a `double` by the compiler in the comparison here?
